feat(ui): if possible, cap window resize at current monitor bounds

Tom94 opened the pull request:
Currently, obtaining the current monitor bounds is limited to

- [x] Wayland
- [x] X11
- [x] Windows
- [x] macOS

Fixes #362

Tom94 committed:
feat(ui): if possible, cap window resize at current monitor bounds

fililip commented on src/ImageViewer.cpp:
```diff
@@ -1713,7 +1713,21 @@ void ImageViewer::resizeToFit(nanogui::Vector2i targetSize) {
+        glfwGetMonitorWorkarea(monitor, &pos.x(), &pos.y(), &size.x(), &size.y());
+        if (glfwGetPlatform() == GLFW_PLATFORM_WAYLAND) {
+            size /= pixel_ratio();
+        }
```
Oh, sorry, I just compiled this branch and it seems that it's unnecessary now, and it actually reduces the max window size to an incorrect value. Removing these lines works correctly on both 1x and 2x scaling

fililip replied:
Oh, that's because `glfwGetMonitorWorkarea` seems to return 1080p for my 4K monitor now... weird

Edit: only for 2x scaling, lower factors (but greater than 1) always return 2160p.

fililip replied:
I think it's a different thing; `pixel_ratio()` is always correct (1, 1.5, 1.75, 2), but `glfwGetMonitorWorkarea` only returns `[1920, 1080]` for 200% scaling.

That division operation was definitely correct (it certainly works with fractional scaling), something seems to be wrong with the way GLFW interprets monitor sizes now. (Did a dependency update change something here?)

Tom94 replied:
At the place where `glfwGetMonitorWorkarea` was previously called, `pixel_ratio()` would always return 1 in the scenario I described above. My suspicion is that KDE sends over correspondingly down-scaled monitor work areas as it sends the correct content scale such that their product always reproduces your true resolution.

If you're up for testing, you could try printing the results of `glfwGetMonitorWorkarea` and `pixel_ratio()` at the top of the `ImageViewer` constructor.

(No dependency update -- I'm already maintaining forks of all the relevant dependencies, so this was an isolated change.)

fililip replied:
I rebased and recompiled the branch, here are my results:

Constructor:

```
20:00:26 INFO     monitor 0 size: [1462999040, 32765]
20:00:26 INFO     monitor 1 size: [3840, 2160]
20:00:26 INFO     pixel_ratio(): 1
```


Primary screen:

```
20:01:06 INFO     monitor 0 size: [-184148240, 32764]
20:01:06 INFO     monitor 1 size: [3840, 2160]
20:01:06 INFO     pixel_ratio(): 2
20:01:06 INFO     current monitor size: [3840, 2160]
```

Secondary screen:

```
20:01:15 INFO     monitor 0 size: [1616729504, 32767]
20:01:15 INFO     monitor 1 size: [3840, 2160]
20:01:15 INFO     pixel_ratio(): 1
20:01:15 INFO     current monitor size: [1920, 1080]
```

This means that now the current monitor call returns the correct screen, but `glfwGetMonitors` appears to add an uninitialized(?) handle as the first element of that pointer array. Accessing the third element yields the secondary screen's dimensions:

```
20:03:45 INFO     monitor 0 size: [-708719424, 32764]
20:03:45 INFO     monitor 1 size: [3840, 2160]
20:03:45 INFO     monitor 2 size: [1920, 1080]
```

Tom94 replied:
Thank you. I'm honestly not sure what happened there. I can't reproduce the issue on my single monitor setup, but I tried another small change as a shot in the dark. 🤞

Tom94 committed:
feat(ui): robustify resizeToFit to Wayland peculiarities

- Handles late DPI scaling / current monitor information
- Handles Wayland compositors that always report as maximized (e.g.
Hyprland)

Tom94 committed:
feat(ui): clamp resized window to monitor on x11, macOS, win32
